### PR TITLE
fix(metrics): Enforce total ordering of all messages

### PR DESCRIPTION
**Pull request by john-z-yang**
### Overview

Enforce total ordering of all valid and invalid messages, so when invalid message is put to the dlq, it is guaranteed to be in the buffer.

Ideally, the validation and filtering of invalid messages should take place in another step. It might be worth considering refactoring this further in the future

### CODE REVIEW

File: src/sentry/sentry_metrics/consumers/indexer/multiprocess.py [hunk 1]
```diff
@@ -38,6 +40,8 @@ def __init__(
         self.__poll_start_time = time.time()
         self.__poll_duration_sum = 0.0
 
+        self.__last_offsets: MutableMapping[Partition, int] = defaultdict(lambda: -1)
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/consumers/indexer/multiprocess.py:43]: nit: This default is pretty unintuitive to me, why does it need to be a defaultdict with a default of -1? Can't a missing key just mean no last offset?

File: src/sentry/sentry_metrics/consumers/indexer/multiprocess.py [hunk 2]
```diff
@@ -68,14 +72,24 @@ def poll(self) -> None:
...
+    def submit(self, message: Message[Union[KafkaPayload, FilteredPayload, None]]) -> None:
         if isinstance(message.payload, FilteredPayload):
             # FilteredPayload will not be commited, this may cause the the indexer to consume
             # and produce invalid message to the DLQ twice if the last messages it consume
             # are invalid and is then shutdown. But it will never produce valid messages
             # twice to snuba
             # TODO: Use the arroyo producer which handles FilteredPayload elegantly
             return
+
+        ((partition, offset),) = message.committable.items()
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/consumers/indexer/multiprocess.py:84]: if committable contains more than one partition/offset are you discarding the others?

File: src/sentry/sentry_metrics/consumers/indexer/multiprocess.py (hunk 1 above)
Comment (lynnagara) [reply src/sentry/sentry_metrics/consumers/indexer/multiprocess.py:43]: why do you need to do all the last offset accounting anyway? aren't they ordered now so we don't need to worry about that?

File: src/sentry/sentry_metrics/consumers/indexer/multiprocess.py [hunk 3]
```diff
@@ -76,6 +76,7 @@ def submit(self, message: Message[Union[KafkaPayload, FilteredPayload]]) -> None
             # twice to snuba
             # TODO: Use the arroyo producer which handles FilteredPayload elegantly
             return
+
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/consumers/indexer/multiprocess.py:79]: nit: better to avoid touching this file at all

**Review by lynnagara — APPROVED** [review]
Just a note that this has a hard dependency on Arroyo 2.14.18 and should not be merged before https://github.com/getsentry/sentry/pull/59654.

Approving in advance though.

File: src/sentry/sentry_metrics/consumers/indexer/multiprocess.py (hunk 3 above)
Comment (john-z-yang) [reply src/sentry/sentry_metrics/consumers/indexer/multiprocess.py]: Ooops, fixed. Thanks!

**Pull request merged**